File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int maxLeft = height[0], maxRight= height[n-1];
        int left = 0, right = n-1;
        int answer=0;
        while(left<right){
            if(maxLeft<maxRight){
                answer+= max(0,maxLeft-height[left]);
                left++;
                maxLeft = max(maxLeft, height[left]);
            }
            else{
                answer += max(0,maxRight-height[right]);
                right--;
                maxRight = max(maxRight, height[right]);
            }
        }
        return answer;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        maxLeft = height[left]
        maxRight = height[right]

        trapped = 0
       
        while(left<right):
            if(maxLeft<maxRight):
                trapped+= maxLeft - height[left]
                left+=1
                maxLeft = max(maxLeft, height[left])
            else: 
                trapped+=maxRight - height[right]
                right-=1
                maxRight = max(maxRight, height[right])

        return trapped


    
     